meraki - Rewrite update requirement check

kbreit opened the pull request:
##### SUMMARY
- Rewrite of the Meraki utility's idempotency check
- Check now operates recursively and works on multiple types
- Order of lists matter
- This will likely change behavior for some modules because there could be a change on whether the data structure is completely overwritten or if it's patched

###### Modules Tested Via Integration Tests
- [x] meraki_admin
- [x] meraki_config_template
- [x] meraki_content_filtering
- [x] meraki_device
- [x] meraki_mr_l3_firewall
- [x] meraki_mx_l3_firewall
- [x] meraki_network
- [x] meraki_organization
- [x] meraki_snmp
- [x] meraki_ssid
- [x] meraki_static_route
- [x] meraki_switchport
- [x] meraki_syslog
- [x] meraki_vlan



Fixes #48279

##### ISSUE TYPE
- Bugfix Pull Request

##### COMPONENT NAME
meraki

##### ANSIBLE VERSION
```
ansible 2.8.0.dev0 (meraki/network_cleanup bb2aca9f4a) last updated 2018/11/08 20:20:56 (GMT -500)
  config file = None
  configured module search path = ['/Users/kbreit/.ansible/plugins/modules', '/usr/share/ansible/plugins/modules']
  ansible python module location = /Users/kbreit/Documents/Programming/ansible/lib/ansible
  executable location = /Users/kbreit/Documents/Programming/ansible/bin/ansible
  python version = 3.5.4 (default, Feb 25 2018, 14:56:02) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]

```

kbreit committed:
Rewrite idempotency check
- Check now operates recursively and works on multiple types
- Order of lists matter

kbreit committed:
Remove blank line for lint

kbreit commented:
@dagwieers I have done some testing on this and there's a corner case which is failing.

```
"original": {"name": "IntTestNetworkWireless","tags": None, "timeZone": "America/Chicago", "type": "wireless"}
"payload": {"name": "IntTestNetworkWireless", "timeZone": "America/Chicago", "type": "wireless"}
```

In this case, the playbook isn't proposing tags but the returned data from Meraki always shows a tags key, even if it's `None`. I can't add the key to `optional_ignore` as it is used sometimes. Where in the code would we do this check for None?

kbreit commented:
@dagwieers Real time updates. I changed the line in `filter_items()` to:

`if k not in self.ignored_keys and v is not None:`

This seems to be working and I'm running tests now.

kbreit committed:
Fixed idempotency checks in meraki_ssid
- New sanitize() method for finding keys unique in compared dicts
- Fixed bug in meraki_ssid where SSID specified by number breaks
  - This will require a backport
- Converted ignored_keys from tuple to list

kbreit committed:
Made changes required for idempotency

kbreit commented:
@dagwieers Take a look at this when you can. It has a lot of required changes.

dagwieers commented on lib/ansible/module_utils/network/meraki/meraki.py:
```diff
@@ -83,7 +84,7 @@ def __init__(self, module, function=None):
                          'admins': '/organizations/{org_id}/admins',
                          'configTemplates': '/organizations/{org_id}/configTemplates',
-                         'samlRoles': '/organizations/{org_id}/samlRoles',
+                         'samlymbols': '/organizations/{org_id}/samlRoles',
```
samlymbols looks like a typo ?

dagwieers reviewed (approved):
LGTM, but I don't see why you would stop testing some failure scenarios.

kbreit commented:
@dagwieers Where am I removing error checks? I generally see that I'm adding some.

dagwieers commented:
@kbreit In the integration tests various tasks are commented out in this PR.

kbreit commented:
@dagwieers Could you take a look at this? I'm running my integration tests again before getting this merged and have a failure. Error is below. I can convert everything to string easily, but what risks are there converting to either unicode for everything or string for everything?

`"Datatypes don't match: <type 'unicode'> vs <type 'str'>"}`

dagwieers commented:
@kbreit I need more info, but the different here is related to Python2 vs Python3 and you most likely need to use to_native() or to_text() in order to cast to the appropriate type.

kbreit committed:
Add changelog fragment

kbreit committed:
Add unidirectional option for testing

kbreit committed:
Disable option 1 check

kbreit commented:
-label: affects_2.8
+label: affects_2.9

kbreit committed:
General fixes for is_update_required testing
- Added commented out debug statements in method
- Fixed ignored_keys modifications

trishnaguha reviewed (commented):
is the commented out code WIP?

kbreit commented:
@trishnaguha No. it’s an alternative algorithm which I will probably remove before I pull WIP and merge.

kbreit committed:
Remove old commented algorithm
